radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/core/reconstruction/BerkeleyRecoverableProcessedTxnStore.java

```diff
@@ -0,0 +1,261 @@
+			var key = new DatabaseEntry(Longs.toByteArray(Long.MAX_VALUE));
+			var value = new DatabaseEntry();
+			cursor.getSearchKeyRange(key, value, null);
+			var status = cursor.getPrev(key, value, null);
+			if (status == SUCCESS) {
+				var accumulatorHash = HashCode.fromBytes(value.getData());
+				var stateVersion = Longs.fromByteArray(key.getData());
+				this.accumulatorState = new AccumulatorState(stateVersion, accumulatorHash);
+			} else {
+				this.accumulatorState = new AccumulatorState(0, HashUtils.zero256());
+			}
+		}
+	}
+
+	@Override
+	public void close() {
+		if (recoverableTransactionsDatabase != null) {
+			recoverableTransactionsDatabase.close();
+		}
+
+		if (accumulatorDatabase != null) {
+			accumulatorDatabase.close();
+		}
+	}
+
+	public Optional<HashCode> getAccumulator(long stateVersion) {
+		if (stateVersion == 0) {
+			return Optional.of(HashUtils.zero256());
+		}
+		var key = new DatabaseEntry(Longs.toByteArray(stateVersion));
+		var value = new DatabaseEntry();
+		var result = accumulatorDatabase.get(null, key, value, null);
+		if (result != SUCCESS) {
+			return Optional.empty();
+		}
+		return Optional.of(HashCode.fromBytes(value.getData()));
+	}
+
+	public List<RecoverableProcessedTxn> get(long index, long limit) {
+		try (var cursor = recoverableTransactionsDatabase.openCursor(null, null)) {
+			var iterator = new Iterator<RecoverableProcessedTxn>() {
+				final DatabaseEntry key = new DatabaseEntry(Longs.toByteArray(index));
+				final DatabaseEntry value = new DatabaseEntry();
+				OperationStatus status = cursor.get(key, value, Get.SEARCH, null) != null ? SUCCESS : OperationStatus.NOTFOUND;
+
+				@Override
+				public boolean hasNext() {
+					return status == SUCCESS;
+				}
+
+				@Override
+				public RecoverableProcessedTxn next() {
+					if (status != SUCCESS) {
+						throw new NoSuchElementException();
+					}
+					RecoverableProcessedTxn next;
+					try {
+						next = serialization.fromDson(Compress.uncompress(value.getData()), RecoverableProcessedTxn.class);
+					} catch (IOException e) {
+						throw new IllegalStateException("Failed to deserialize committed transaction.", e);
+					}
+
+					status = cursor.getNext(key, value, null);
+					return next;
+				}
+			};
+			return Streams.stream(iterator)
+				.limit(limit)
+				.onClose(cursor::close)
+				.toList();
+		}
+	}
+
+	@Override
+	public void process(Transaction dbTxn, REProcessedTxn txn, long stateVersion, Function<SystemMapKey, Optional<RawSubstateBytes>> mapper) {
+		if (accumulatorState.getStateVersion() != stateVersion - 1) {
+			throw new IllegalStateException("Accumulator out of sync.");
+		}
+
+		var nextAccumulatorState = ledgerAccumulator.accumulate(accumulatorState, txn.getTxnId().asHashCode());
+		txn.stateUpdates()
+			.filter(u -> u.getParsed() instanceof RoundData)
+			.map(u -> (RoundData) u.getParsed())
+			.filter(r -> r.getTimestamp() > 0)
+			.map(RoundData::asInstant)
+			.forEach(timestamp::set);
```

Comment:
do we want the latest timestamp here? are `stateUpdates` already sorted (or always have the same value)? if no, then perhaps it should be something like:
```
			.max(Instant::compareTo)
			.ifPresent(timestamp::set);

```

Reply:
StateUpdates are sorted. And timestamp isn't guaranteed to be increasing so we must associate each transaction with the timestamp in the given round.

Reply:
I have some logic to make a "better" normalised timestamp in the gateway service